glusterd2/commands/volumes/brick-replace.go

```diff
@@ -0,0 +1,221 @@
+	"github.com/gluster/glusterd2/glusterd2/bricksplanner"
+	"github.com/gluster/glusterd2/glusterd2/gdctx"
+	"github.com/gluster/glusterd2/glusterd2/peer"
+	restutils "github.com/gluster/glusterd2/glusterd2/servers/rest/utils"
+	"github.com/gluster/glusterd2/glusterd2/transaction"
+	"github.com/gluster/glusterd2/glusterd2/volume"
+	"github.com/gluster/glusterd2/pkg/api"
+	gderrors "github.com/gluster/glusterd2/pkg/errors"
+
+	"github.com/gorilla/mux"
+	"github.com/pborman/uuid"
+)
+
+func registerReplaceBrickStepFuncs() {
+	var sfs = []struct {
+		name string
+		sf   transaction.StepFunc
+	}{
+		{"brick-replace.PrepareBricks", prepareBricks},
+		{"brick-replace.ReplaceVolinfo", replaceVolinfo},
+		{"brick-replace.StartBrick", startBrick},
+	}
+	for _, sf := range sfs {
+		transaction.RegisterStepFunc(sf.sf, sf.name)
+	}
+}
+
+func replaceBrickHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	logger := gdctx.GetReqLogger(ctx)
+	volname := mux.Vars(r)["volname"]
+
+	if !volume.IsValidName(volname) {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, gderrors.ErrInvalidVolName)
+		return
+	}
+
+	// Get Volume Info
+	vol, err := volume.GetVolume(volname)
+	if err != nil {
+		status, err := restutils.ErrToStatusCode(err)
+		restutils.SendHTTPError(ctx, w, status, err)
+		return
+	}
+
+	// Unmarshal ReplaceBrickReq
+	var req api.ReplaceBrickReq
+	if err := restutils.UnmarshalRequest(r, &req); err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, gderrors.ErrJSONParsingFailed)
+		return
+	}
+
+	if uuid.Parse(req.SrcPeerID) == nil {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, "Invalid peerID passed in url")
+		return
+	}
+
+	if err := validateVolumeFlags(req.Flags); err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, err)
+		return
+	}
+
+	subVols := vol.Subvols
+
+	var srcBrickInfo brick.Brickinfo
+	subVolIndex := 0
+	brickIndex := 0
+LOOP:
+	for index := range subVols {
+		for i, brick := range subVols[index].Bricks {
+			// Get Source Brick Info
+			if brick.PeerID.String() == req.SrcPeerID && brick.Path == req.SrcBrickPath {
+				subVolIndex = index
+				brickIndex = i
+				srcBrickInfo = brick
+				break LOOP
+			}
+		}
+	}
+
+	excludeZones := make([]string, 0)
+	for svIndex, sv := range subVols {
+		// if SubvolZonesOverlap is true then bricks of only that particular
+		// subvolume will be considered.
+		if req.SubvolZonesOverlap && subVolIndex != svIndex {
+			continue
+		}
+		for _, b := range sv.Bricks {
+			p, err := peer.GetPeer(b.PeerID.String())
+			if err != nil {
+				restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+				return
+			}
+			excludeZones = append(excludeZones, p.Metadata["_zone"])
+		}
+	}
+	req.ExcludeZones = append(req.ExcludeZones, excludeZones...)
+
+	subvolumes := make([]api.SubvolReq, 0)
+	volreq := api.VolCreateReq{
+		Subvols:      subvolumes,
+		Size:         vol.Capacity,
+		LimitPeers:   req.LimitPeers,
```

Comment:
have we done validation for `LimitPeers` does this peer id really exists?
same for `LimitZones` and `ExcludePeers`

Reply:
separate validation is not be required. If user gives invalid list then it will not affect the functionality.

Reply:
if the request succeeds, the user will be an assumption that request is valid.

Reply:
ok, then this needs to be done during volume create as well

Reply:
ok, I will add validation for both volume_create and replace brick in a separate PR.

Reply:
please  create an issue, so we won't forget it